File: typescript/packages/core/src/evals/live.ts
import type { Agent } from "../agent/agent";
import { liveRunOf } from "../agent/registry";
import type { RunResult } from "../agent/result";
import type { CaseDir } from "./case-dir";
import type { CaseLog } from "./checks";
import { JUDGE_V1, transcript } from "./judge";
import { runJudge } from "./judge-run";
import {
  answerOf,
  costOf,
  type Env,
  EVAL_PRINCIPAL,
  eventsOf,
  guarded,
  type LiveOutcome,
  NO_CALLS,
  unfinished,
} from "./live-env";
import { simulateCase } from "./simulate";
import { stubQueue } from "./stub-queue";
import { treeTotals } from "./tree-totals";

// The live check (spec lane 22, C): the current agent answers the case input on a new thread,
// with every mediated call answered from the recorded stubs (an eval never performs a real side
// effect), and a judge model grades the whole turn against the rubric. A case with `simulate`
// goes to the multi-turn conversation instead (spec lane 32, B).

export type { Env, Live, LiveOutcome } from "./live-env";
export { EVAL_PRINCIPAL } from "./live-env";

async function judged(
  c: CaseDir,
  criteria: readonly string[],
  lead: Extract<RunResult<unknown>, { status: "completed" }>,
  env: Env,
): Promise<LiveOutcome> {
  const events = await eventsOf(env.store, lead.thread);
  const answer = answerOf(lead.output);
  const items = transcript(events);
  const ran = await runJudge(
    {
      instructions: JUDGE_V1,
      task: c.meta.input?.text ?? "",
      events,
      answer,
      rubric: criteria,
    },
    env,
  );
  if (ran.kind === "blocked") return { kind: "blocked", model: ran.model };
  const calls = {
    agent: (await treeTotals(env.store, lead.thread)).requests,
    user: 0,
    judge: ran.calls,
  };
  const cost = await costOf(
    ran.thread === undefined ? [lead.thread] : [lead.thread, ran.thread],
  );
  if (ran.kind === "error")
    return { kind: "error", reason: ran.reason, calls, cost };
  const passed = ran.graded.filter((v) => v.pass).length;
  return {
    kind: "graded",
    check: {
      answer,
      transcript_items: items.length,
      score: passed / ran.graded.length,
      fresh_sandbox: env.fresh,
      verdicts: [...ran.graded],
      ...(env.kept
        ? { thread_id: lead.thread.id, judge_thread_id: ran.thread.id }
        : {}),
    },
    calls,
    cost,
  };
}

/** Runs the case input on the current agent and grades it. */
export async function liveCheck(
  c: CaseDir,
  log: CaseLog,
  target: Agent<never, unknown>,
  env: Env,
): Promise<LiveOutcome> {
  const criteria = [...(c.meta.rubric ?? []), ...(env.live.rubric ?? [])];
  if (criteria.length === 0)
    return {
      kind: "skipped",
      reason: "no_rubric",
      calls: NO_CALLS,
      cost: null,
    };
  if (c.meta.simulate !== undefined)
    return simulateCase(c, log, target, criteria, c.meta.simulate, env);
  const text = c.meta.input?.text;
  const run = liveRunOf(target);
  if (text === undefined || run === undefined)
    return {
      kind: "skipped",
      reason: "offline_not_runnable:content_input",
      calls: NO_CALLS,
      cost: null,
    };
  const lead = await guarded(() =>
    run(text, {
      store: env.store,
      principal: EVAL_PRINCIPAL,
      budget: env.live.budget,
      stub: stubQueue(c.stubs ?? { stubs: [] }, "turn"),
    }),
  );
  if ("blocked" in lead) return { kind: "blocked", model: lead.blocked };
  if ("config" in lead)
    return { kind: "error", reason: lead.config, calls: NO_CALLS, cost: null };
  if (lead.status !== "completed") {
    const calls = {
      agent: (await treeTotals(env.store, lead.thread)).requests,
      user: 0,
      judge: 0,
    };
    return {
      kind: "error",
      reason: unfinished(lead),
      calls,
      cost: await costOf([lead.thread]),
    };
  }
  return judged(c, criteria, lead, env);
}

File: python/src/threads/evals/live.py
"""The live check (spec lane 22, C): the current agent answers the case input on a new thread, with
every mediated call answered from the recorded stubs (an eval never performs a real side effect),
and a judge model grades the whole turn against the rubric. A case with `simulate` goes to the
multi-turn conversation instead (spec lane 32, B)."""

from collections.abc import Sequence

from pydantic import JsonValue

from threads.agents.results import Completed, Thread
from threads.evals.case_dir import CaseDir
from threads.evals.checks import CaseLog
from threads.evals.judge import JUDGE_V1, Graded, transcript
from threads.evals.judge_run import JudgeAsk, run_judge
from threads.evals.live_env import (
    EVAL_PRINCIPAL,
    Blocked,
    Env,
    EvalAgent,
    Live,
    Misconfigured,
    Outcome,
    answer_of,
    cost_sum,
    events_of,
    new_thread,
    run_agent,
    unfinished,
)
from threads.evals.simulate import simulate_case
from threads.evals.stub_queue import stub_queue
from threads.evals.tree_totals import tree_totals
from threads.log import Event

__all__ = ["EVAL_PRINCIPAL", "Env", "EvalAgent", "Live", "Outcome", "live_check"]


def _check(
    graded: Sequence[Graded], answer: JsonValue, events: Sequence[Event], env: Env
) -> dict[str, JsonValue]:
    passed = sum(1 for g in graded if g.passed)
    return {
        "answer": answer,
        "transcript_items": len(transcript(events)),
        "score": passed / len(graded),
        "fresh_sandbox": env.fresh,
        "verdicts": [g.to_json() for g in graded],
    }


async def _judged(
    case: CaseDir, target: EvalAgent, lead: Completed[str], criteria: Sequence[str], env: Env
) -> Outcome:
    events = await events_of(env.store, lead.thread.branch)
    answer = answer_of(target.definition, lead.output)
    text = case.meta.input.text if case.meta.input is not None else None
    ask = JudgeAsk(JUDGE_V1, text or "", events, answer, criteria)
    ran = await run_judge(ask, env)
    agent_calls = (await tree_totals(env.store, lead.thread.branch)).requests
    if ran.kind == "blocked":
        return Outcome("blocked", reason=ran.reason)
    threads: list[Thread] = [lead.thread]
    if ran.thread is not None:
        threads.append(ran.thread)
    cost = await cost_sum(threads)
    if ran.kind == "error":
        return Outcome(
            "error",
            reason=ran.reason,
            agent_calls=agent_calls,
            judge_calls=ran.calls,
            cost=cost,
        )
    check = _check(ran.graded, answer, events, env)
    if env.kept and ran.thread is not None:
        check |= {"thread_id": lead.thread.id, "judge_thread_id": ran.thread.id}
    return Outcome("graded", check, agent_calls=agent_calls, judge_calls=ran.calls, cost=cost)


async def _one_turn(case: CaseDir, target: EvalAgent, text: str, env: Env) -> Outcome:
    """Lane 22's single graded turn: one run on a new thread, then the judge."""
    criteria = [*(case.meta.rubric or ()), *env.live.rubric]
    script: dict[str, JsonValue] = case.stubs if isinstance(case.stubs, dict) else {"stubs": []}
    thread = await new_thread(env)
    lead = await run_agent(target.definition, text, env, thread, stubs=stub_queue(script, "turn"))
    if isinstance(lead, Blocked):
        return Outcome("blocked", reason=lead.model)
    if isinstance(lead, Misconfigured):
        return Outcome("error", reason=lead.reason)
    if not isinstance(lead, Completed):
        calls = (await tree_totals(env.store, lead.thread.branch)).requests
        return Outcome(
            "error",
            reason=unfinished(lead),
            agent_calls=calls,
            cost=await cost_sum([lead.thread]),
        )
    return await _judged(case, target, lead, criteria, env)


async def live_check(case: CaseDir, log: CaseLog, target: EvalAgent, env: Env) -> Outcome:
    """Runs the case input on the current agent and grades it."""
    criteria = [*(case.meta.rubric or ()), *env.live.rubric]
    if not criteria:
        return Outcome("skipped", reason="no_rubric")
    if case.meta.simulate is not None:
        return await simulate_case(case, log, target, criteria, case.meta.simulate, env)
    text = case.meta.input.text if case.meta.input is not None else None
    if text is None:
        return Outcome("skipped", reason="offline_not_runnable:content_input")
    return await _one_turn(case, target, text, env)
